pkg/sources/transformer/grpc_transformer.go

```diff
@@ -0,0 +1,123 @@
+	functionsdk "github.com/numaproj/numaflow-go/pkg/function"
+	"github.com/numaproj/numaflow-go/pkg/function/client"
+	"github.com/numaproj/numaflow/pkg/forward/applier"
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/udf/function"
+
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/emptypb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// gRPCBasedTransformer applies user defined transformer over gRPC (over Unix Domain Socket) client/server where server is the transformer.
+type gRPCBasedTransformer struct {
+	client functionsdk.Client
+}
+
+var _ applier.MapApplier = (*gRPCBasedTransformer)(nil)
+
+// NewGRPCBasedTransformer returns a new gRPCBasedTransformer object.
+func NewGRPCBasedTransformer() (*gRPCBasedTransformer, error) {
+	c, err := client.New()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new gRPC client: %w", err)
+	}
+	return &gRPCBasedTransformer{c}, nil
+}
+
+// NewGRPCBasedTransformerWithClient need this for testing
+func NewGRPCBasedTransformerWithClient(client functionsdk.Client) *gRPCBasedTransformer {
+	return &gRPCBasedTransformer{client: client}
+}
+
+// CloseConn closes the gRPC client connection.
+func (u *gRPCBasedTransformer) CloseConn(ctx context.Context) error {
+	return u.client.CloseConn(ctx)
+}
+
+// WaitUntilReady waits until the client is connected.
+func (u *gRPCBasedTransformer) WaitUntilReady(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("failed on readiness check: %w", ctx.Err())
+		default:
+			if _, err := u.client.IsReady(ctx, &emptypb.Empty{}); err == nil {
+				return nil
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+func (u *gRPCBasedTransformer) ApplyMap(ctx context.Context, readMessage *isb.ReadMessage) ([]*isb.Message, error) {
+	key := readMessage.Key
+	payload := readMessage.Body.Payload
+	offset := readMessage.ReadOffset
+	parentPaneInfo := readMessage.PaneInfo
+	var d = &functionpb.Datum{
+		Key:       key,
+		Value:     payload,
+		EventTime: &functionpb.EventTime{EventTime: timestamppb.New(parentPaneInfo.EventTime)},
+		Watermark: &functionpb.Watermark{Watermark: timestamppb.New(readMessage.Watermark)},
+	}
+
+	ctx = metadata.NewOutgoingContext(ctx, metadata.New(map[string]string{functionsdk.DatumKey: key}))
+	// TODO - change to MapTFn once numaflow-go sdk changes merge in.
+	datumList, err := u.client.MapFn(ctx, d)
+	if err != nil {
+		return nil, function.ApplyUDFErr{
+			UserUDFErr: false,
+			Message:    fmt.Sprintf("gRPC client.MapFn failed, %s", err),
+			InternalErr: function.InternalErr{
+				Flag:        true,
+				MainCarDown: false,
+			},
+		}
+	}
+
+	writeMessages := make([]*isb.Message, 0)
+	for i, datum := range datumList {
+		key := datum.Key
+		if datum.EventTime != nil {
+			// Transformer supports changing event time.
+			parentPaneInfo.EventTime = datum.EventTime.EventTime.AsTime()
+		}
```

Comment:
shouldn't there be an `else` condition so that if we get a `nil` from the source, we can assign the default value we got from the source? Basically

```
EventTime: parentPaneInfo.EventTime),
Watermark: readMessage.Watermark),
```

or do we handle that in the SDK side?

Reply:
Firstly we guarantee non-nil because mapT requires user to specify an event time. Secondly even if nil, the following code (line 112) will pass the original parentPaneInfo to the `writeMessage`.

Reply:
Some thoughts:

1. The current `MessageToxxx` [helper functions](https://github.com/numaproj/numaflow-go/blob/main/pkg/function/message_t.go#L16-L28) in the SDK all require `eventTime`, shall we allow the users to only return a `MessageT` without giving `eventTime`, which means they don't care about it, and use the original eventTime is okay.
2. Thinking of following scenarios, are we expecting these behaviors - if a UDF image is given to `transformer`, nothing is broken, the transformer just works as a regular UDF without having the ability to set eventTime; If a transformer image is give to UDF, it still works as a regular UDF, even though it sets the eventTime in the return, but it will be ignored. 

@vigith @KeranYang

Reply:
sounds like a good idea to me.

Reply:
+1. Thinking about how to implement this.
1. On SDK side, we can unify `Message` and `MessageT` into one generic type such that transformer and udf share exact same contract and can replace each other from user spec.
2. On numaflow side, transformer gRPC will check event time and update `writeMessage` while udf gRPC just don't check event time and pass on the parentPaneInfo.

I am gonna create a separate issue to track this change.

Reply:
https://github.com/numaproj/numaflow/issues/504